### ISSUE TITLE
Rustyline crashes in linux if you hold enter

### ISSUE THREAD

**Issue by jonathandturner** [kkawakam/rustyline#291]
Repro:

```
> cargo run --example example
1> [now hold enter here]
```

Once you hold enter, after a few seconds you'll see:

```
1> 
Line: 
2> 
Line: 
3> 
Line: 
4> 
Line: 
5> 
Line: 

Error: Io(Kind(InvalidData))
```

**Comment by gwenn** [comment]
Sorry,
I cannot reproduce.
* x86_64-apple-darwin / rustc 1.40.0-nightly (c27f7568b 2019-10-13) (macOS Catalina)
* x86_64-unknown-linux-gnu / rustc 1.38.0 (625451e37 2019-09-23) (archlinux VM)

**Comment by jonathandturner** [comment]
My platform:

```
Linux pop-os 5.0.0-25-generic #26-Ubuntu SMP Thu Aug 1 12:04:58 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
rustc 1.39.0-nightly (521d78407 2019-08-25)
```

**Comment by gwenn** [comment]
I've also tried rustc on linux but still no error:
I guess that the VM/VirtualBox layer prevents me to reproduce the bug.

**Comment by jonathandturner** [comment]
That could be it. My guess is that it comes from the speed that the enter is happening, so trying it on a native install of Linux might be the best way to try to repro.

**Comment by Detegr** [comment]
@jonathandturner which shell are you using? My machine is native Arch Linux `Linux 5.3.5-arch1-1-ARCH #1 SMP PREEMPT Mon Oct 7 19:03:08 UTC 2019 x86_64 GNU/Linux`, compiled with `rustc 1.39.0-nightly (9eae1fc0e 2019-08-23)` and I'm unable to repro this in Alacritty or xterm. I tried setting key repeat to an absurdly high value with `xset r rate 200 250` but it does not have any effect.

**Comment by jonathandturner** [comment]
I can't repro in alacritty, but I can repro inside the default GNOME terminal for Pop! OS (an Ubuntu-based distro)

**Comment by gwenn** [comment]
@jonathandturner Would you mind trying to generate a backtrace (`RUST_BACKTRACE=1`) ?

**Comment by jonathandturner** [comment]
@gwenn - it panics without a backtrace, but experimenting a bit, I see that it's originating here:

https://github.com/kkawakam/rustyline/blob/39f35781de120865f51290f877f8a4292ac594b7/src/tty/unix.rs#L689

**Comment by jonathandturner** [comment]
With a bit more experimenting: I understand that this is verifying that you should be properly reset, but commenting it out doesn't seem to show any issues of not getting reset correctly:

https://github.com/kkawakam/rustyline/blob/39f35781de120865f51290f877f8a4292ac594b7/src/tty/unix.rs#L688-L699

I was wondering what kinds of issues this was trying to protect from. Maybe there's another way to do it?

**Comment by gwenn** [comment]
My wild guess is that we receive `Enter`/'\r' instead of `Esc`/`\x1b` because the key press happens after:
https://github.com/kkawakam/rustyline/blob/39f35781de120865f51290f877f8a4292ac594b7/src/tty/unix.rs#L677
But before:
https://github.com/kkawakam/rustyline/blob/39f35781de120865f51290f877f8a4292ac594b7/src/tty/unix.rs#L682

We can still ignore unexpected/user input in this case.

**Comment by gwenn** [comment]
See #292

**Comment by gwenn** [comment]
`Liner` seems to always write "⏎":
https://github.com/redox-os/liner/blob/master/src/editor.rs#L156

**Comment by jonathandturner** [comment]
@gwenn - I tried out #292 and it fixes the issue for me.

**Comment by gwenn** [comment]
Fix by release 5.0.4
